File: SearchUnknownsizeArray.java
// Time Complexity : O(logn)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : NO -(dont have premium leetcode)
// Any problem you faced while coding this : no

// Your code here along with comments explaining your approach
import javax.lang.model.util.ElementScanner6;

public class SearchUnknownsizeArray {
    public int search(ArrayReader reader, int target) {
        int high = 1, low = 0, mid = -1;
        while (reader.get(high) != Integer.MAX_VALUE) {
            high = high * 2;
        }

        while (low <= high) {
            mid = low + (high - low) / 2;
            if (reader.get(mid) == target) {
                return mid;
            } else if (reader.get(mid) < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return -1;
    }
}

File: SearchUnknownSizeArray.py
# Time Complexity : O(log n)
# Space Complexity : O(log n)
# Did this code successfully run on Leetcode : YES
# """
# This is ArrayReader's API interface.
# You should not implement it, or speculate about its implementation
# """
# class ArrayReader:
#    def get(self, index: int) -> int:

class Solution:
    def search(self, reader: 'ArrayReader', target: int) -> int:
        #check if target is first element itself or not
        if reader.get(0) == target:
            return 0

        #initialize low and high
        low = 0
        high = 1

        #increase value of high as per as requirement
        while reader.get(high) <= target:
            low = high
            high = high * 2

        #simple binary search
        while low <= high:
            mid = (low + high) // 2
            value = reader.get(mid)

            if value == target:
                return mid
            elif target < value:
                high = mid - 1
            else:
                low = mid + 1

        return -1
